README.md
# employee-management-api

## Api crud built in Flask


## *Getting started!*

### Clonar el repo

    git clone git@github.com:jeremyakd/employee-management-api.git

    cd employee-management-api

# Metodo de deploy by docker-compose
El repo cuenta con un compose con imagenes ya pusheadas a mi dockerhub.
Instalar docker & docker-compose.
- [Instalar docker](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-20-04-es)
- [Instalar docker-compose](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-compose-on-ubuntu-20-04-es
)
## Run pulleando las imagenes
    docker-compose -f docker-compose.yml up -d

## Run buildeando las imagenes
    docker-compose -f docker-compose-build.yml up -d --build

# Run artesanal
### Instalar dependencias 
<small>Se entiende que se cuenta con virtualenv ya instalado. De lo contrario visitar => https://j2logo.com/virtualenv-pip-librerias-python/</small>

    virtualenv env --python=python3
    source env/bin/activate
    pip insttall -r requirements.txt

### Run

    flask run

### Test

    nosetests --verbose

### Documentacio Swagger
:link: http://localhost:5000/swagger/ :link:

<hr>
<br>

### Comando para obtener graficos.

- <small>En terminal python</small>:

        python scripts/graphics.py

- <small>Ejecutado con docker-compose</small>:

        docker-compose exec employee-management-api python /app/scripts/graphics.py

### Modelo de datos de la API

**empleado**
- Id
- Nombre
- Apellido
- DNI
- Fecha_Nacimiento
- Fecha_Ingreso

## Build docker image
    docker build --tag employee-management-api:1.0.0 .

## Run on docker image
    docker run --publish 5000:5000 employee-management-api:1.0.0

changelog.md
# Changelog de cambios.

- 25-11-2021 => *@jeremyakd*    
\- Creacion de estructura de la API

- 26-11-2021 => *@jeremyakd*  
\- Se agrega blbliteca de mongo para gurdar los datos en una DB.  
\- Se crea la lógica de las operaciones de CRUD

- 27-11-2021 => *@jeremyakd*  
\- Se agrega planilla con datos dummie. Se actualiza campos a tipo datetime para filtrar. Se crean funciones de filtro por edad y por antiguedad.
\- Se agrega docker-compose con mongo por problemas con la db local. Se agrega archivo para llenar la DB con datos de la libreria Faker.

- 28-11-2021 => *@jeremyakd*  
\- Se agrega archivo de Graficos, Se actualiza script para cargar datos Dummie.
\- Se agrega uri para crear y borrar datos de prueba.
\- Se agrega Dockerfile, wsgy, script para ejecutar gunicorn para prueba de build.
\- Update requirements.txt
\- Se agrega configuracion para que se conecte a mongo

- 29-11-2021 => *@jeremyakd*  
\- Se agrega variables de entorno, Se agrega archivo para levantar mongo solo, Se actualiza Dockerfile.

- 30-11-2021 => *@jeremyakd*  
\- Se actualiza el acceso a las variables de entornos en los demas archivos.
\- Se agrega documentacion con Swagger.
\- Se actualiza el acceso a las variables de entornos en el archivo que faltaba.
\- Se actualiza requirements.txt.
\- Se agregan variables de entonro, Se crean algunos casos de prueba.
\- Se actualiza requirements.txt.

- 01-12-2021 => *@jeremyakd*  
\- Se actualiza tipo de respuesta.
\- Limpieza de código, orden del repo.
\- Reestructuración del repositorio, actualizacion del README.
\- Correcciones varias minimas.
\- Se agrega servidor Nginx y un compose con build.
\- Actualizacion de changelog
\- Actualizacion de docker-compose con imagenes de mi hub.
\- Limpieza de código
\- Se actualiza documentación
\- Se agrega descripción a las funciones de los endpoints.

test/app_test.py
"""
The tests to run in this project.
To run the tests type,
$ nosetests --verbose
"""

from nose.tools import assert_true
import requests, os

BASE_URL = os.environ.get('BASE_URL', 'http://localhost:5000')

class NewUUID():
    """
    The new uuid created when creating a new employee.
    The NewUUID is used for further tests.
    """
    def __init__(self, value):
        self.value = value


def test_get_individual_request_404():
    "Test getting a non existent request"
    response = requests.get('%s/employee/an_incorrect_id' % (BASE_URL))
    assert_true(response.status_code == 404)


def test_add_new_employee():
    "Test adding a new employee"
    payload = {
        'nombre': 'Joe',
        'apellido': 'Doe',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    NewUUID.value = str(response.json()['id'])
    assert_true(response.status_code == 201)


def test_get_new_employee():
    "Test getting the new employee"
    url = '%s/employee/%s' % (BASE_URL, NewUUID.value)
    response = requests.get(url)
    assert_true(response.ok)


def test_edit_new_employee_nombre():
    "Test editing the new employee name"
    payload = {
        'nombre': 'Juan',
        'apellido': 'Doe',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee/%s' %
                            (BASE_URL, NewUUID.value), json=payload)
    assert_true(response.json()['nombre'] == "Juan")


def test_edit_new_employee_apellido():
    "Test editing the new employee apellido"
    payload = {
        'nombre': 'Joe',
        'apellido': 'Perez',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee/%s' %
                            (BASE_URL, NewUUID.value), json=payload)
    assert_true(response.json()['apellido'] == "Perez")


def test_add_new_employee_bad_name_key():
    "Test adding a new employee with a bad name key"
    payload = {
        'bad_key': 'Joe',
        'apellido': 'Doe',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    assert_true(response.status_code == 400)


def test_add_new_employee_bad_date_format():
    "Test adding a new employee with a bad date format"
    payload = {
        'nombre': 'Joe' ,
        'apellido': 'Doe',
        'dni': '11223344',
        'fecha_nacimiento': 888888,
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    assert_true(response.status_code == 400)


def test_add_new_employee_no_name_key():
    "Test adding a new employee without name"
    payload = {
        'apellido': 'Doe',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    assert_true(response.status_code == 400)


def test_add_new_employee_no_apellido_key():
    "Test adding a new employee without apellido"
    payload = {
        'nombre': 'Joe',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    assert_true(response.status_code == 400)


def test_add_new_employee_unicode_name():
    "Test adding a new employee with a unicode title"
    payload = {
        'nombre': '▚Ⓜ⌇⇲',
        'apellido': 'Doe',
        'dni': '11223344',
        'fecha_nacimiento': '13-09-1984',
        'fecha_ingreso': '20-03-2010'
    }
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    assert_true(response.ok)


def test_add_new_employee_no_payload():
    "Test adding a new employee with no payload"
    payload = None
    response = requests.post('%s/employee' % (BASE_URL), json=payload)
    assert_true(response.status_code == 400)


def test_delete_new_employee():
    "Test deleting the new employee"
    response = requests.delete('%s/employee/%s' % (BASE_URL, NewUUID.value))
    assert_true(response.status_code == 204)


def test_delete_new_employee_404():
    "Test deleting the new employee that was already deleted"
    response = requests.delete('%s/employee/%s' % (BASE_URL, NewUUID.value))
    assert_true(response.status_code == 404)


scripts/graphics.py
import mongoengine
import pandas as pd
import matplotlib.pyplot as plt
import plotext
import pymongo
import termplotlib as tpl
import os

collection = False

MONGO_URL = os.environ.get("MONGO_URL", 'localhost')
MONGO_PORT = os.environ.get("MONGO_PORT", '27017')
MONGO_DB = os.environ.get("MONGO_DB", 'employee')

def conect_mongo():
    """
    Connect to DB and return the db and collections object.
    """
    client = mongoengine.connect(host='mongodb://' + MONGO_URL + ':' +  MONGO_PORT + '/' + MONGO_DB)
    db = client.employee
    collection = db.employee
    data = pd.DataFrame(list(collection.find()))
    return data, collection


def extract_data(data, column):
    """
    Extrae de la DB las colunmas enviadas por parametro y
    devuelve una lista de el año actual menos la fecha de la columna.
    """
    col_list = data[column].tolist()
    return [ ( 2021 - d.to_pydatetime().year ) for d in col_list ]

def get_quantity(ages):
    """
    Esta funcion recibe una lista de edades y las fitra
    por rango etario.
    """
    q = [0,0,0,0,0]
    for a in ages:
        if a < 29:
            q[0]+=1
        elif a < 39:
            q[1]+=1
        elif a < 49:
            q[2]+=1
        elif a < 59:
            q[3]+=1
        else:
            q[4]+=1
    return q

def get_employees_dates_and_antiquty(collections):
    """
    Recibe una coleccion y devuelve una lista con 2 elementos.
    - Una lista con las edades.
    - Una lista con la antigüedad.
    La lista devuelta está ordenada por edad de forma ascendente.
    """
    np_array = [[],[]]
    for c in collections.find().sort('fecha_nacimiento',pymongo.DESCENDING):
        np_array[0].append(2021 - c['fecha_ingreso'].year)
        np_array[1].append(2021 - c['fecha_nacimiento'].year)
    return np_array


def print_graphic_age_bar(ages):
    """
    Recibe una lista de datetimes, Las filtra
    y luego imprime en consola un grafico de barras relacionando
    edad y entigüedad.
    """
    plotext.bar(['20-29', '30-39', '40-49', '50-59', '60-69'], get_quantity(ages))
    plotext.title("Personas por rango etario")
    plotext.show()

def get_graphic_age(ages):
    """
    Recibe una lista de datetimes, Las filtra
    y luego muestra un grafico de barras relacionando
    edad y entigüedad.
    """
    fig, ax = plt.subplots()
    ax.bar(['20-29', '30-39', '40-49', '50-59', '60-69'], get_quantity(ages))
    plt.title("Personas por rango etario")
    plt.show()



def print_graphic_age(entrys):
    """
    Recibe una lista de ordenadas de 2 listas de valores.
    Luego imprime en consola un grafico relacionando edad
    y antiguedad en la empresa.
    """
    fig = tpl.figure()
    fig.plot(entrys[1], entrys[0], width=150, height=20)
    fig.show()

def get_graphic_antiquity(entrys):
    """
    Recibe una lista de ordenadas de 2 listas de valores
    para dibujar un grafico en consola.
    """
    fig, ax = plt.subplots()
    ax.plot(entrys[1], entrys[0])
    plt.show()

mongo_data, collection = conect_mongo()
ages_and_antiquities = get_employees_dates_and_antiquty(collection)

edades = extract_data( mongo_data, 'fecha_nacimiento', )
ingresos = extract_data( mongo_data, 'fecha_ingreso' )


# Gráfico que permita visualizar a los empleados por rango etario
print_graphic_age_bar(edades)
# -------------------------------------------------------------------------
# Gráfico que relacione edad con antigüedad en la empresa 
print_graphic_age(ages_and_antiquities)
# -------------------------------------------------------------------------

get_graphic_age( edades )
get_graphic_antiquity(ages_and_antiquities)




scripts/load_data.py
from faker import Factory
import mongoengine
import datetime
import os

MONGO_URL = os.environ.get("MONGO_URL", 'localhost')
MONGO_PORT = os.environ.get("MONGO_PORT", '27017')
MONGO_DB = os.environ.get("MONGO_DB", 'employee')

def create_names(cant):
    client = mongoengine.connect(host='mongodb://' + MONGO_URL + ':' +  MONGO_PORT + '/' + MONGO_DB)
    db = client.employee
    fake = Factory.create()
    for x in range(int(cant)):
        fake_name = fake.first_name()
        fake_last_name = fake.last_name()
        fake_dni = str(fake.pyint(min_value=1000000, max_value=99999999))
        fake_f_nacimiento = fake.date_between_dates(
            datetime.datetime.strptime('1961-01-01', '%Y-%m-%d'),
            datetime.datetime.strptime('2003-01-01', '%Y-%m-%d'),
        )
        fake_f_ingreso = fake.date_between_dates(
            datetime.datetime.strptime('2011-01-01', '%Y-%m-%d')
        )
        try:
            result = db.employee.insert_one(
                {
                'nombre': fake_name,
                'apellido': fake_last_name,
                'dni': fake_dni,
                'fecha_nacimiento': datetime.datetime.combine(fake_f_nacimiento, datetime.datetime.min.time()),
                'fecha_ingreso': datetime.datetime.combine(fake_f_ingreso, datetime.datetime.min.time()),
                }
            )
        except Exception as e:
            print("e", e)
            return False
    return "Se cargaron {} empleados.".format(cant)



app.py
from flask import Flask, jsonify, request

from bson import ObjectId
from flask_pymongo import PyMongo

from dateutil.relativedelta import relativedelta
import datetime, os

from flask_swagger_ui import get_swaggerui_blueprint

MONGO_URL = os.environ.get("MONGO_URL", 'localhost')
MONGO_PORT = os.environ.get("MONGO_PORT", '27017')
MONGO_DB = os.environ.get("MONGO_DB", 'employee')

app = Flask(__name__)

app.config['MONGO_URI'] = 'mongodb://' + MONGO_URL + ':' +  MONGO_PORT + '/' + MONGO_DB

mongo = PyMongo(app)

db = mongo.db


### swagger specific ###
SWAGGER_URL = '/swagger'
API_URL = '/static/swagger.json'
SWAGGERUI_BLUEPRINT = get_swaggerui_blueprint(
    SWAGGER_URL,
    API_URL,
    config={
        'app_name': "Register employees"
    }
)
app.register_blueprint(SWAGGERUI_BLUEPRINT, url_prefix=SWAGGER_URL)
### end swagger specific ###


# ROUTES 

######################################################## ALTA ########################################################
@app.route('/employee', methods=['POST'])
def createEmployee():
    """
    Create a new employee with data provided
    """
    id = None
    try:
        id = db.employee.insert({
            'nombre': request.json['nombre'],
            'apellido': request.json['apellido'],
            'dni': request.json['dni'],
            'fecha_nacimiento': datetime.datetime.strptime(request.json['fecha_nacimiento'], '%d-%m-%Y'),
            'fecha_ingreso': datetime.datetime.strptime(request.json['fecha_ingreso'], '%d-%m-%Y'),
        })
    except Exception as e:
        return "Error al crear", 400
    #return jsonify(str(ObjectId(id)))
    return jsonify({
        "message":"Empleado creado",
        "id": str(ObjectId(id)),
    }), 201

######################################################## BAJA ########################################################
@app.route('/employee/<id>', methods=['DELETE'])
def deleteEmployee(id):
    """
    Delete a employee with id provided
    """
    try:
        employee = db.employee.find_one_and_delete({'_id': ObjectId(id)})
        if not employee:
            return "No se encontró empleado.", 404
    except Exception as e:
        return "No se encontró empleado.", 404
    return jsonify({
        "message":"Employee Deleted",
        "id": str(ObjectId(id)),
    }), 204
	# return jsonify({'message': 'Employee Deleted'})

#################################################### MODIFICACION ####################################################
@app.route('/employee/<id>', methods=['POST'])
def updateEmployee(id):
    """
    Update a employee with id provided
    """
    db.employee.update_one({'_id': ObjectId(id)}, {"$set": {
		'nombre': request.json['nombre'],
		'apellido': request.json['apellido'],
		'dni': request.json['dni'],
		'fecha_nacimiento': request.json['fecha_nacimiento'],
		'fecha_ingreso': request.json['fecha_ingreso'],
	}})
    employee = db.employee.find_one({'_id': ObjectId(id)})
    return jsonify({
        '_id': str(ObjectId(employee['_id'])),
        'nombre': employee['nombre'],
        'apellido': employee['apellido'],
        'dni': employee['dni'],
        'fecha_nacimiento': employee['fecha_nacimiento'],
        'fecha_ingreso': employee['fecha_ingreso'],
    })


#################################################### CONSULTA POR ID #################################################
@app.route('/employee/<id>', methods=['GET'])
def getEmployee(id):
    """
    Get a employee with id provided
    """
    employee = None
    try:
        employee = db.employee.find_one({'_id': ObjectId(id)})
    except Exception as e:
        print("e", e)
    if employee is not None:
        return jsonify({
            '_id': str(ObjectId(employee['_id'])),
            'nombre': employee['nombre'],
            'apellido': employee['apellido'],
            'dni': employee['dni'],
            'fecha_nacimiento': employee['fecha_nacimiento'],
            'fecha_ingreso': employee['fecha_ingreso'],
            })
    else:
        return "No se encontró empleado.", 404


#################################################### FILTRO POR EDAD #################################################
@app.route('/filter_by_age/<edad>', methods=['GET'])
def employees_by_age(edad):
    """
    Return a list of employees filtered age provided.
    """
    now = datetime.datetime.now()
    # con la fecha seteo el dia limite de busqueda y lo guardo en la variable hasta
    hasta = now - relativedelta(years=int(edad))
    # le resto un año y un dia para setear el umbral de fechas de busqueda
    desde = (hasta - relativedelta(years=1) - datetime.timedelta(days=1)) # bajo lte
    employee_filtered = db.employee.find({"fecha_nacimiento": { "$gte": desde, "$lte": hasta }} )
    employees = []
    for doc in employee_filtered:
        employees.append({
            '_id': str(ObjectId(doc['_id'])),
            'nombre': doc['nombre'],
            'apellido': doc['apellido'],
            'dni': doc['dni'],
            'fecha_nacimiento': doc['fecha_nacimiento'],
			'fecha_ingreso': doc['fecha_ingreso'],
        })
    response = {
        'message': 'Se encontraron {} empleados con {} años.'.format(len(employees), edad),
        'response': employees
    }
    return jsonify(response)


################################################ FILTRO POR ANTIGUEDAD ###############################################
@app.route('/filter_by_antiquity/<antiguedad>', methods=['GET'])
def employees_by_antiquity(antiguedad):
    """
    Return a list of employees filtered antiquity provided.
    """
    now = datetime.datetime.now()
    hasta = now - relativedelta(years=int(antiguedad))
    desde = (hasta - relativedelta(years=1) - datetime.timedelta(days=1))
    employee_filtered = db.employee.find({"fecha_ingreso": { "$gte": desde, "$lte": hasta }} )
    employees = []
    for doc in employee_filtered:
        employees.append({
            '_id': str(ObjectId(doc['_id'])),
            'nombre': doc['nombre'],
            'apellido': doc['apellido'],
            'dni': doc['dni'],
            'fecha_nacimiento': doc['fecha_nacimiento'],
			'fecha_ingreso': doc['fecha_ingreso'],
        })
    return jsonify(employees)


############################################### TODOS LOS EMPLEADOS ##################################################
@app.route('/employees', methods=['GET'])
def getEmployees():
    """
    Return all employees on DB.
    """
    employees = []
    for doc in db.employee.find():
        employees.append({
            '_id': str(ObjectId(doc['_id'])),
            'nombre': doc['nombre'],
            'apellido': doc['apellido'],
            'dni': doc['dni'],
            'fecha_nacimiento': doc['fecha_nacimiento'],
			'fecha_ingreso': doc['fecha_ingreso'],
        })
    return jsonify(employees)

############################################ BORRAR TODOS LOS EMPLEADOS #############################################
@app.route('/employees', methods=['DELETE'])
def deleteEmployees():
    """
    Delete all employees on DB.
    PLEASE BE CAREFUL !!!
    """
    db.employee.delete_many({})
    return jsonify({"message": "droped!"})


################################################ CARGAR EMPLEADOS ###################################################
@app.route('/load-data/<cantidad>', methods=['POST'])
def load_data(cantidad):
    """
    Receives an amount per parameter and load on DB with fake data.
    """
    from scripts.load_data import create_names
    _employees = create_names(cantidad)
    return jsonify({ "mesage": _employees })

if __name__ == "__main__":
    app.run(debug=True)
